connector/src/main/java/com/google/cloud/spark/bigquery/ProtoBufUtils.java

```diff
@@ -0,0 +1,380 @@
+
+        DynamicMessage.Builder messageBuilder = DynamicMessage.newBuilder(schemaDescriptor);
+        int i = 1;
+        while(i <= schemaDescriptor.getFields().size()) {
+            StructField sparkField = schema.fields()[i-1];
+            DataType sparkType = sparkField.dataType();
+            if (sparkType instanceof StructType) {
+                messageBuilder.setField(schemaDescriptor.findFieldByNumber(i),
+                        createSingleRowMessage((StructType)sparkType,
+                                schemaDescriptor.findNestedTypeByName(RESERVED_NESTED_TYPE_NAME +i),
+                                (InternalRow)row.get(i-1, sparkType)));
+            }
+            else {
+                messageBuilder.setField(schemaDescriptor.findFieldByNumber(i),
+                        convert(sparkField,
+                                row.get(i-1, sparkField.dataType())));
+            }
+            i++;
+        }
+
+        return messageBuilder.build();
+    }
+
+    @VisibleForTesting
+    protected static Descriptors.Descriptor toDescriptor (StructType schema)
+            throws Descriptors.DescriptorValidationException {
+        DescriptorProtos.DescriptorProto.Builder descriptorBuilder = DescriptorProtos.DescriptorProto.newBuilder()
+                .setName("Schema");
+
+        StructField[] fields = schema.fields();
+
+        DescriptorProtos.DescriptorProto descriptorProto = buildDescriptorProtoWithFields(descriptorBuilder, fields, 0);
+
+        return createDescriptorFromProto(descriptorProto);
+    }
+
+    @VisibleForTesting
+    protected static Object convert (StructField sparkField, Object sparkValue) {
+        if (sparkValue == null) {
+            if (!sparkField.nullable()) {
+                throw new IllegalArgumentException("Non-nullable field was null.");
+            }
+            else {
+                return null;
+            }
+        }
+
+        DataType fieldType = sparkField.dataType();
+
+        if (fieldType instanceof ArrayType) {
+            ArrayType arrayType = (ArrayType)fieldType;
+            boolean containsNull = arrayType.containsNull(); // elements can be null.
+            DataType elementType = arrayType.elementType();
+
+            ArrayData arrayData = (ArrayData)sparkValue;
+            Object[] sparkValues = arrayData.toObjectArray(elementType);
+
+            return Arrays.stream(sparkValues).map(value -> {
+                if (!containsNull && value == null) {
+                    throw new IllegalArgumentException("Encountered a null value inside a non-null-containing array.");
+                }
+                else {
+                    return toAtomicProtoRowValue(elementType, value);
+                }
+            } ).collect(Collectors.toList());
+        } else if (fieldType instanceof StructType) {
+            throw new IllegalStateException("Method did not expect to convert a StructType instance.");
```

Comment:
It is supported, it's just the conversion is done in line 210, outside of this method, in createSingleRowMessage():
`if (sparkType instanceof StructType) {
                messageBuilder.setField(schemaDescriptor.findFieldByNumber(i),
                        createSingleRowMessage((StructType)sparkType,
                                schemaDescriptor.findNestedTypeByName(RESERVED_NESTED_TYPE_NAME +i),
                                (InternalRow)row.get(i-1, sparkType)));
            }
            else {`
Should I simply delete line 254? Since method convert() is never called if fieldType = StructType.